### PR TITLE
[PXM-20] [PXM-25] [PXM-32]

### PR DESCRIPTION
**Pull request by 0xjohnnycagewins**

### CODE REVIEW

File: contracts/Genesis.sol [hunk 1]
```diff
@@ -33,40 +37,33 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+    uint256 public constant DEMI_GODS_MAX_SUPPLY = 400;
+    uint256 public constant ELEMENTALS_MAX_SUPPLY = 550;
     bool public mintActive = false;
     string public baseTokenURI;
-    Counters.Counter private nextTokenId;
+    Counters.Counter private tokenCounter;
+    Counters.Counter private randomNumberRequestsCounter;
+    Counters.Counter private godsCounter;
+    Counters.Counter private demiGodsCounter;
+    Counters.Counter private elementalsCounter;
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:51]: why did you change this?
I was thinking having 1 variable might be more efficient. Are counters cheap?

File: contracts/Genesis.sol [hunk 2]
```diff
@@ -33,40 +37,33 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
     uint256 public constant PRICE = 0.0000001 ether;
-    uint256 public constant MAX_PER_MINT = 1;
+    uint256 public constant WHITELIST_MINT_COUNT = 1;
+    uint256 public constant GODS_MAX_SUPPLY = 50;
+    uint256 public constant DEMI_GODS_MAX_SUPPLY = 400;
+    uint256 public constant ELEMENTALS_MAX_SUPPLY = 550;
     bool public mintActive = false;
     string public baseTokenURI;
-    Counters.Counter private nextTokenId;
+    Counters.Counter private tokenCounter;
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:47]: the reason I had called it `nextTokenId` is because `counter` is a bit counter-intuitive cause it's not really a counter since we always use it for the next id and not the current one.

File: contracts/Genesis.sol [hunk 3]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+    }
 
-    function setMerkleTreeRoot(bytes32 _merkleTreeRoot) public onlyOwner {
-        merkleTreeRoot = _merkleTreeRoot;
+    /**
+     * Mint a god
+     * @param nftOwner address of the NFT owner
+     * @param tokenId id of the token to be minted
+     */
+    function mintGod(address nftOwner, uint256 tokenId) private {
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:116]: I would use the syntax `to` instead of `nftOwner` here. He is not really the owner at this point, but he will be.

File: contracts/Genesis.sol [hunk 4]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+        godsCounter.increment();
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:117]: Why is there no `require` here?
There should be one to check the `totalSupply` and one to make sure there are still god available.

File: contracts/Genesis.sol [hunk 5]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
     /**
-     * Disables mint active state
+     * Mint a token
+     * @param nftOwner address of the NFT owner
+     * @param tokenId id of the token to be minted
+     * @param randomNumber random number returned by Chainlink
      */
-    function disableMint() public onlyOwner {
-        mintActive = false;
+    function mint(address nftOwner, uint256 tokenId, uint256 randomNumber) private {
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:152]: see other comment for `nftOwner`, I don't really like that naming.

File: contracts/Genesis.sol [hunk 6]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+        tokenCounter.increment();
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:153]: Ok, I understand why there are no `require`, because these are private and called from a method that do have the require?
Let me know how you feel, I'm wondering if we shouldn't make sure every function in itself has the `require` it needs, but it might get repetitive. I'll let you decide on this.

File: contracts/Genesis.sol [hunk 7]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+        addressToMintCount[nftOwner]++;
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:154]: Add a comment here because I don't like this so much but it makes sense in this specific case, a comment would make it clear we understand there could be overflow but wouldn't be cause of the limitations we have

File: contracts/Genesis.sol [hunk 8]
```diff
@@ -82,23 +79,80 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
-     * Enables mint
+     * Mint a demi-god
+     * @param nftOwner address of the NFT owner
+     * @param tokenId id of the token to be minted
      */
-    function enableMint() public onlyOwner {
-        mintActive = true;
+    function mintDemiGod(address nftOwner, uint256 tokenId) private {
+        demiGodsCounter.increment();
+        tokenIdToTraits[tokenId] = TokenTraits(TokenType.DEMI_GOD);
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:129]: `DEMI_GOD` are generative and have random traits too.

File: contracts/Genesis.sol [hunk 9]
```diff
@@ -107,55 +161,60 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
      * @param randomNumber random number returned by Chainlink
      */
     function fulfillRandomness(bytes32 requestId, uint256 randomNumber) internal override {
+        randomNumberRequestsCounter.decrement();
         address nftOwner = requestIdToSender[requestId];
-        uint256 tokenId = requestIdToTokenId[requestId];
-        require(nextTokenId.current() > tokenId, "TokenId has not been minted yet!");
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:112]: Why did you remove that require? I'm fairly certain we should keep it for security purposes. This can't really be called except from ChainLink but still.

File: contracts/Genesis.sol [hunk 10]
```diff
@@ -107,55 +161,60 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
     }
 
     /**
-     * Mint a given number of tokens
-     * The sender has to be included in the free mint list (except if they are the owner)
+     * Requests minting
+     * @param sender sender of the request 
+     */
+    function requestMint(address sender) private {
+        addressToRequestCount[sender]++;
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:183]: Again, I understand in theory it wouldn't overflow, but I don't think we should use this here, especially cause we never make checks against it to ensure it never is above 1.

File: contracts/Genesis.sol (hunk 3 above)
Comment (GabrielCartier) [reply contracts/Genesis.sol:116]: or `sender` to be consistent.

File: contracts/Genesis.sol [hunk 11]
```diff
@@ -107,55 +161,60 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
     }
 
     /**
      * Requests a random number to Chainlink VRF for a given token id
+     * @param sender sender of the request 
      * @param tokenId id of the token
      * @return requestId id of the request sent to Chainlink
      */
-    function requestRandomNumberForTokenId(uint256 tokenId) internal returns (bytes32 requestId) {
+    function requestRandomNumberForTokenId(address sender, uint256 tokenId) internal returns (bytes32 requestId) {
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:222]: why send it when it's there? You're just duplicating a parameter here.

File: contracts/Genesis.sol [hunk 12]
```diff
@@ -107,55 +161,60 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+        mint(nftOwner, requestIdToTokenId[requestId], randomNumber);
     }
 
     /**
-     * Mint a given number of tokens
-     * The sender has to be included in the free mint list (except if they are the owner)
+     * Requests minting
+     * @param sender sender of the request 
+     */
+    function requestMint(address sender) private {
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:182]: same thing here, you have this through the `msg` param.

File: contracts/Genesis.sol [hunk 13]
```diff
@@ -171,38 +230,46 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
-            elementalsCount = elementalsCount.sub(1);
             return TokenType.ELEMENTAL;
         }
     }
 
+    /**
+    * Returns the total supply
+    * @return count count of NFTs minted so far
+    */
+    function totalSupply() public view returns (uint256 count) {
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:277]: shouldn't be `pure`?

File: test/test-genesis.ts [hunk 14]
```diff
@@ -50,49 +72,51 @@ describe('Genesis Contract', () => {
...
       ),
-    ).to.be.revertedWith('Minting is not active');
-    await contract.connect(owner).enableMint();
-    expect(await contract.mintActive()).to.be.true;
-    await contract.connect(owner).disableMint();
-    expect(await contract.mintActive()).to.be.false;
+    ).to.be.revertedWith('Pausable: paused');
+    await contract.connect(owner).unpause();
+    expect(await contract.paused()).to.be.false;
+    await contract.connect(owner).pause();
```
Comment (GabrielCartier) [inline test/test-genesis.ts:96]: weird test here? why pause and unpause?
Perhaps this should be it's own test?

File: contracts/Genesis.sol [hunk 15]
```diff
@@ -171,38 +230,46 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
...
+     * Verifies the proof of the sender to confirm they are in given list
      * @param nonce nonce to be used
+     * @param root Merkle tree root
      * @param proof proof
      * @return valid TRUE if the proof is valid, FALSE otherwise
      */
-    function verifyProof(uint256 nonce, bytes32[] memory proof) internal view returns (bool valid) {
-        return MerkleProof.verify(proof, merkleTreeRoot, generateLeaf(nonce, msg.sender));
+    function verifyProof(uint256 nonce, bytes32 root, bytes32[] memory proof) internal view returns (bool valid) {
+        return MerkleProof.verify(proof, root, generateLeaf(nonce, msg.sender));
```
Comment (GabrielCartier) [inline contracts/Genesis.sol:248]: One thing I noticed we aren't doing is checking if the `root` is actually set. What if it isn't?
Doesn't seem like it will fail, we need to test that and perhaps add a `require` here to check that the root isn't `0`?

File: test/test-genesis.ts [hunk 16]
```diff
@@ -101,38 +125,56 @@ describe('Genesis Contract', () => {
...
+      .connect(owner)
+      .setWhiteListMerkleTreeRoot(whiteListMerkleTreeRoot);
+    await expect(
+      contract
+        .connect(whitelisted)
+        .mintWhitelist(whiteListNonce, whiteListProof, {
+          value: ethers.utils.parseEther('0.000000000001'),
+        }),
+    ).to.be.revertedWith('Not enough ETH');
+  });
```
Comment (GabrielCartier) [inline test/test-genesis.ts:166]: I had added this test on my side too, great minds think alike huh?

File: test/test-utils.ts [hunk 17]
```diff
@@ -95,3 +95,39 @@ export const mint = async (
...
+export const freeMint = async (
+  count: number,
+  minter: SignerWithAddress,
+  nonce: number,
+  proof: string[],
+  contract: Contract,
+  oracle: SignerWithAddress,
+  coordinatorMockAddress: string,
+) => {
+  const mintTx = await contract.connect(minter).freeMint(count, nonce, proof);
```
Comment (GabrielCartier) [inline test/test-utils.ts:116]: Or have a flag `freeMint` and change the method.
MINOR anyway.

**Review by GabrielCartier — CHANGES_REQUESTED** [review]
I think there are some stuff to add in there. 
You are a bit lax on the require, I understand why they might not really be required for internal functions but I'm wondering if we shouldn't make sure they all can run independently without running into issues.

Either way, some stuff needs to be added IMO.

Good job!

File: contracts/Genesis.sol (hunk 9 above)
Comment (0xjohnnycagewins) [reply contracts/Genesis.sol:112]: added `require` in other functions
